src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -410,267 +584,458 @@ private string GenerateForObject(TypeMetadata typeMetadata)
+
+            private string GetEarlyNullCheckSource(bool canBeNull)
+            {
+                return canBeNull
+                    ? $@"if ({ValueVarName} == null)
+        {{
+            {WriterVarName}.WriteNullValue();
+            return;
+        }}
+"
+                    : null;
+            }
+
+            private string GetSerializeLogicForNonPrimitiveType(string typeInfoPropertyName, string valueToWrite, bool serializationLogicGenerated)
+            {
+                string typeInfoRef = $"{_currentContext.ContextTypeRef}.Default.{typeInfoPropertyName}";
+
+                if (serializationLogicGenerated)
+                {
+                    return $"{typeInfoPropertyName}{SerializeMethodNameSuffix}({WriterVarName}, {valueToWrite});";
+                }
+
+                return $"{JsonSerializerTypeRef}.Serialize({WriterVarName}, {valueToWrite}, {typeInfoRef});";
+            }
+
+            private enum DefaultCheckType
             {
-                string typeCompilableName = typeMetadata.CompilableName;
-                string typeFriendlyName = typeMetadata.FriendlyName;
+                None,
+                Null,
+                Default,
+            }
 
-                return @$"{GetUsingStatementsString(typeMetadata)}
+            private string WrapSerializationLogicInDefaultCheckIfRequired(string serializationLogic, string propValue, DefaultCheckType defaultCheckType)
+            {
+                if (defaultCheckType == DefaultCheckType.None)
+                {
+                    return serializationLogic;
+                }
 
-namespace {_generationNamespace}
+                string defaultLiteral = defaultCheckType == DefaultCheckType.Null ? "null" : "default";
+                return $@"
+        if ({propValue} != {defaultLiteral})
+        {{{serializationLogic}
+        }}";
+            }
+
+            private string[] GetRuntimePropNames(List<PropertyGenerationSpec>? properties, JsonKnownNamingPolicy namingPolicy)
+            {
+                if (properties == null)
+                {
+                    return Array.Empty<string>();
+                }
+
+                int propCount = properties.Count;
+                string[] runtimePropNames = new string[propCount];
+
+                // Compute JsonEncodedText values to represent each property name. This gives the best throughput performance
+                for (int i = 0; i < propCount; i++)
+                {
+                    PropertyGenerationSpec propertySpec = properties[i];
+
+                    string propName = DetermineRuntimePropName(propertySpec.ClrName, propertySpec.JsonPropertyName, namingPolicy);
+                    Debug.Assert(propName != null);
+
+                    runtimePropNames[i] = propName;
+                }
+
+                return runtimePropNames;
+            }
+
+            private string DetermineRuntimePropName(string clrPropName, string? jsonPropName, JsonKnownNamingPolicy namingPolicy)
```

Comment:
When can `jsonPropName` be null? It looks like it defaults to the CLR name elsewhere.

Reply:
Yeah, I had a stray [LOC](https://github.com/dotnet/runtime/blob/1dd5d734dd66d163b48ed406bc71fc308c63e2eb/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L472) that appeared to default the value, but it was a no-op (now deleted). It is `null` when `[JsonPropertyName]` is not specified for the property.